### ISSUE TITLE
Shift+Drag-and-Drop from Dolphin doesn't work on Wayland

### ISSUE THREAD

**Issue by CounterPillow** [mpv-player/mpv#9201]
### Important Information

Provide following Information:
- mpv version: git master
- Linux Distribution and Version: Arch Linux, 5.13.13
- Source of the mpv binary: mpv-build
- Window Manager and version: kwin + plasma on Wayland, Plasma 5.22.5 on Frameworks 5.85
- GPU driver and version: mesa 21.2.1, amdgpu

### Reproduction steps

1. Use KDE on Wayland
2. Play a file in mpv
3. Shift+drag a second file from Dolphin onto mpv

### Expected behavior

Dropped file should be queued up at the end of the playlist

### Actual behavior

Looks like the shift part of dragging and dropping isn't registering at all, and we're just replacing the currently playing file

### Log file

https://0x0.st/-w0h.log

This is probably an issue with kwin somehow not sending us the Shift modifier key in this case, but I'm just reporting it here first to make sure that this isn't some Wayland lolextensions protocol issue.

**Comment by Dudemanguy** [comment]
The same thing happens on sway. I think this is actually on Dolphin to implement. It looks like Dolphin always sends mpv `WL_DATA_DEVICE_MANAGER_DND_ACTION_COPY` which results in replacing the current file. If it sent `WL_DATA_DEVICE_MANAGER_DND_ACTION_MOVE`, that should result an append according to the code. Maybe it's hidden in the settings somewhere. I didn't look too hard or try other file managers.

**Comment by CounterPillow** [comment]
Left a bug report for KDE: https://bugs.kde.org/show_bug.cgi?id=442232

**Comment by Dudemanguy** [comment]
As an aside, the compositor is actually allowed to change the dnd action on the fly according to the protocol which I think should work. So it would not require any code changes from the source application (Dolphin in this case). I'm not sure if anyone currently implements anything like this though.

>Compositors may also change the selected action on the fly, mainly in response to keyboard modifier changes during the drag-and-drop operation.

https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_data_offer

**Comment by lunik1** [comment]
I see the same behaviour on KDE in Chromium and Firefox, could be related or needs its own big report?

**Comment by Dudemanguy** [comment]
There's three parts to drag and drop. The source of the DnD (so Dolphin in this case), what receives it (mpv), and finally the actual compositor needs to implement the protocol as well for this all to work. So if any part of this is not correct, then DnD will not be correct. tl;dr maybe it's the same bug or maybe it's not.

**Comment by appetrosyan** [comment]
Is there any way to override this behaviour?

**Comment by Dudemanguy** [comment]
Is it not fixed actually? The linked issue seems to point to QT which has actually merged a patch related to this.

**Comment by CounterPillow** [comment]
It works if you press shift before dragging and dropping a file, so it's related to https://bugs.kde.org/show_bug.cgi?id=423031

**Comment by CounterPillow** [comment]
Just to give a small update, because KDE Plasma is a buggy shitheap they actually managed to make this worse and now not even holding shift before dragging works. Either way, closing because it's not mpv's problem, but a problem in the incredibly poorly written pile of C++ that is KDE Plasma. What a joke of a fucking desktop environment, but at least it's not GN\*ME.

**Comment by appetrosyan** [comment]
> Just to give a small update, because KDE Plasma is a buggy shitheap they actually managed to make this worse and now not even holding shift before dragging works.

Charming. 

> Either way, closing because it's not mpv's problem, but a problem in the incredibly poorly written pile of C++ that is KDE Plasma.

1. You should close this as not planned. 
2. You shouldn't close this, because `mpv` can fix this problem. Namely, add a config parameter that allows you to trigger the "add" functionality without holding shift. 
3. Just because it's a problem in another project, doesn't mean it's a good look for your project to **wontfix** an issue that can well stay open and just be closed when the issue *in the other project* is fixed. 
4. Evidently, the issue is of agreeing on an API, because KDE's Dolphin seems to be perfectly capable of parsing Shift-drag as separate from drag. 
5. This issue is not exclusive to KDE, and there are projects other than `mpv` that seem to resolve it, suggesting that `mpv` is at least partly culpable. 
6. Denigrating other projects (which massive in scope, contain multiple voluntary contributions and don't have the luxury of regular maintenance) is not a good look for your project. Yes, KDE's code base is problematic. Any project of that size would have such problems. A project of that size evidently is in high demand. Behave like a grown person, and treat your users with some respect.

**Comment by 6r1d** [comment]
> Evidently, the issue is of agreeing on an API, because KDE's Dolphin seems to be perfectly capable of parsing Shift-drag as separate from drag.

I'm surprised it's that complex in a native program, while it is pretty simple to detect in modern browsers with JS.

**Comment by CounterPillow** [comment]
I also can't get a Dolphin-to-Dolphin drag-and-drop to recognise any modifier keys now, so it's 100% something broken on KDE's side and not a matter of a disagreement over protocol or whatever.

**Comment by appetrosyan** [comment]
> I also can't get a Dolphin-to-Dolphin drag-and-drop to recognise any modifier keys now, so it's 100% something broken on KDE's side and not a matter of a disagreement over protocol or whatever.

Given the direction of the conversation, I'd hesitate to say anything is 100%. There are variables for which you cannot control, variables which might take time to discover, and variables which when made apparent might be embarrassing. 

Still I'll err on your side, and pretend that I cannot do Dolphin to Dolphin drag and drop. But instead invite the question of why can't you make drag and drop without the modifier do the same thing? Make it a config parameter, give us a workaround? 

Additionally, while I understand the occasional frustration, and that some large projects have made baffling decisions.  Is this form of antagonism necessary?

**Comment by Dudemanguy** [comment]
Adding a hack to workaround something that should just be fixed in the other application isn't really appealing.

**Comment by lunik1** [comment]
I'd prefer non-shift as append in any case. There's value in making this behaviour configurable outside any workarounds, imho.

**Comment by appetrosyan** [comment]
> Adding a hack to workaround something that should just be fixed in the other application isn't really appealing.

I prefer to think of it as a continuation of the modular, customisable API design. Having the ability to rebind `space` is not in any way different from rebinding `drag'n'drop`.

**Comment by appetrosyan** [comment]
Should we open a separate issue and mark this as "has workaround" citing the new functionality?

**Comment by Dudemanguy** [comment]
>Having the ability to rebind space is not in any way different from rebinding drag'n'drop.

That's not possible on wayland. There's no way to know what modifier is pressed.

**Comment by appetrosyan** [comment]
> That's not possible on wayland. There's no way to know what modifier is pressed.

Maybe I need to paraphrase. 

I can bind Space to quit mpv if I want to. Similarly, I would like to also be able to swap the action that would be triggered on shift drag with the action that is triggered on the normal drag’n’drop. That is more than possible because it does not depend on Wayland, it’s strictly based on exposing a config param that changes the behaviour of a regular drag’ event.

**Comment by Dudemanguy** [comment]
>Similarly, I would like to also be able to swap the action that would be triggered on shift drag with the action that is triggered on the normal drag’n’drop.

mpv doesn't have any knowledge of what keys are being pressed or anything like that. It just knows it received data, that's it. The only thing that could theoretically be done is adding a setting that always overrides whatever action we get from the compositor and always use that.

**Comment by appetrosyan** [comment]
We don’t seem to be understanding each other. 

I understand that the compositor might send different actions instead of raw modifiers + drag’n’drop event. I don’t mind that due to bugs, not all events can be triggered. This is true, and **irrelevant**. 

I want an option inside MPV that changes how MPV interprets the event that is supported by compositors. 

Namely I want to replace the action of killing my playlist and starting from the first entry, with the action that appends the files to the playlist, when I do a regular drag and drop. 

MPV doesn’t need to know what keys are pressed. It needs to receive the files and append them to the playlist, without starting to play the first file, or clearing the playlist.

**Comment by garoto** [comment]
> It needs to receive the files and append them to the playlist, without starting to play the first file, or clearing the playlist.

That's exactly what Dudemanguy has being saying for the past 10hrs or so.

**Comment by Dudemanguy** [comment]
>I want an option inside MPV that changes how MPV interprets the event that is supported by compositors.

That would be the global option that ignores the event action like I said.

**Comment by appetrosyan** [comment]
Is this something the team would be willing to entertain?

**Comment by Dudemanguy** [comment]
Sure doesn't sound horrible.

**Comment by appetrosyan** [comment]
@garoto 

for the record, 
1. it’s closer to three hours, 
2. the only reason I went on to elaborate was because of the following statement, 
> Adding a hack to workaround something that should just be fixed in the other application isn't really appealing.

To which I replied that being able to change the functions triggered by inputs is in no way a hack 

3. and to 

> That's not possible on wayland. There's no way to know what modifier is pressed.

That what I request is not specific to wayland, and indeed something which was suggested by @Dudemanguy themselves. 

4. I wish no further antagonism, and as I said previously, would happily do the work if a PR is accepted.
